\documentclass[11pt,a4paper]{report}

\usepackage[utf8]{inputenc}
\usepackage{graphics}
\usepackage{amsmath}

\usepackage[hidelinks]{hyperref}
\usepackage{url} 

\usepackage{pdfpages}
\usepackage{pdflscape}
\usepackage{pythonhighlight}
\usepackage{amssymb}
\usepackage{braket}
\usepackage{pdfpages}
\usepackage[margin=0.7cm]{caption}

\usepackage[resetlabels,labeled]{multibib}

\newcounter{counter}[section]
\renewcommand{\thecounter}{\thechapter.\arabic{counter}}

\newenvironment{definition}[1]{\bigskip\refstepcounter{counter}\noindent\textbf{Definition \thecounter } \emph{#1} \par\nopagebreak\noindent \begin{itshape}}{\end{itshape}\bigskip}

\newenvironment{example}[1]{\bigskip\refstepcounter{counter}\noindent\textbf{Example \thecounter} \emph{#1} \par\nopagebreak}{\bigskip}

\newenvironment{notation}{\bigskip\refstepcounter{counter}\noindent\textbf{Notation \thecounter~}\nopagebreak \begin{itshape}}{\end{itshape}\bigskip}

\newenvironment{theorem}{\bigskip\refstepcounter{counter}\noindent\textbf{Theorem \thecounter }\nopagebreak \begin{itshape}}{\end{itshape}\medskip}

\newenvironment{lemma}{\bigskip\refstepcounter{counter}\noindent\textbf{Lemma \thecounter }\nopagebreak \begin{itshape}}{\end{itshape}\bigskip}

\newenvironment{remark}{\bigskip\refstepcounter{counter}\noindent\textbf{Remark \thecounter~ }\nopagebreak \begin{itshape}}{\end{itshape}\bigskip}

\definecolor{gray}{rgb}{0.4,0.4,0.4}
\definecolor{graybright}{rgb}{0.5,0.5,0.5}
\definecolor{darkblue}{rgb}{0.0,0.0,0.6}
\definecolor{cyan}{rgb}{0.0,0.6,0.6}

\lstset{
  basicstyle=\ttfamily,
  columns=fullflexible,
  showstringspaces=false,
  commentstyle=\color{gray}\upshape
}

\lstdefinelanguage{XML}
{
  morestring=[b]",
  morestring=[s]{>}{<},
  morecomment=[s]{<?}{?>},
  stringstyle=\color{black},
  identifierstyle=\color{darkblue},
  keywordstyle=\color{cyan},
  morekeywords={xmlns,version,type}% list your attributes here
}

\lstdefinelanguage{LBS}
{
  stringstyle=\color{black},
  keywordstyle=\textbf,
  morekeywords={spec, new, comp, inside}% list your attributes here
}

\begin{document}

%---------preamble generated from fithesis template
\includepdf[pages={2}]{preamble/pre.pdf}

\pagestyle{plain}
\pagenumbering{roman}

\tableofcontents
\newpage

\setcounter{page}{0}
\pagenumbering{arabic}

%--------------------------------------------------------

\chapter{Introduction} \label{chap:intro}

Modelling complex systems in systems biology has to be conducted at several levels of abstraction that reflect well the known information~\cite{Kitano}. At every level, the system has to be described rigorously in a formal language that allows to avoid
misunderstanding and ambiguous interpretations. The more complex the system is, the harder it is to describe it rigorously while not losing human-readability and compactness of the description at the same time.

Traditional approaches used to describe biochemical systems are a~\emph{chemistry approach} employing ``mechanical'' descriptions by chemical reactions or a \emph{mathematical approach} using ordinary differential equations or other mathematical formalisms. An advantage of chemical reactions over mathematical equations is the fact that chemical reactions are composable to some extent, human-readable and well-understood across disciplines while still having executable semantics. 
Moreover, there are methods automatising the generation of mathematical models from chemical equations.

The problem of both approaches is the \emph{scalability} which is understood at two
different levels: scalability of the model description (avoiding combinatorial explosion at the syntactic level) and scalability of the model execution (avoiding combinatorial explosion at the semantics level). Even when the formulation of a model does not run into syntactic scalability problems, the execution or simulation might be infeasible~\cite{Romers107136}. 

The so-called \emph{computational approach}~\cite{Cardelli,Henzinger} offers scalable model description by abstracting the information about individual model components and interactions. The scalability at the semantics level still remains to be a challenge. A promising computational approach is provided by \emph{rule-based modelling}~\cite{kappa_formal,harris2016bionetgen} and process-algebraic frameworks~\cite{Cardelli,BioPEPA,BioSPI}. Rule-based models make a natural extension of the mechanical reaction-based models used in chemistry. Instead of operating with \emph{objects}, rule-based frameworks operate with \emph{types} that allow to avoid the combinatorial explosion that occurs when underlying objects are specified directly. The semantics of the model is given in terms of \emph{rules} defined on given types. The rules express a high degree of modularity which helps to avoid the explicit enumeration of all possible molecular species or all the states of a system. They also enable a more natural starting point for model development than making \emph{ad hoc} assumptions to decide the model scope. An important advantage of the rule-based approach is that mathematical models can be automatically generated from it. In particular, instead of relying on a single mathematical formalism, different mathematical models can thus be obtained for a given model (e.g., ODEs~\cite{camporesi2017k}, PDEs~\cite{andrews2016smoldyn}, chemical master equation or continuous-time Markov chains~\cite{pauleve2010generic,sneddon2011efficient}, reaction-diffusion systems~\cite{sorokina2013simulator}, etc.). 

The long-term aim of our research is the development of Comprehensive Modelling Platform~\cite{klement2013comprehensive,BCS} -- a general modelling framework that combines model building, model analysis, and annotation tasks in a single public site related to a single system. It respects the need for maintaining existing ODE models (which is still a typical scenario in systems biology) but allows to align them with a mechanistic rule-based description that is understandable by biologists, compact in size, and executable in terms of allowing basic analysis tasks ensuring consistency of the description. Such a comprehensive solution allows to support modellers effort in building mathematical models that have clear biochemical meaning and can be
easily integrated. Moreover, mechanistic descriptions can be later used as standalone computational models having all advantages of rule-based modelling. 

To that end, we have pioneered an idea of combining advantages of rule-based modelling with the simplicity of chemical reactions by introducing a rule-based language called \emph{Biochemical Space Language} (BCSL)~\cite{dved2016formal}. It has the following key aspects: \emph{human-readability} -- easy to read, write, and maintain;
\emph{executability} -- formal executable semantics is defined allowing efficient static analysis and consistency checking; \emph{universality} -- principally different cellular processes can be sufficiently combined in a single specification; and \emph{syntactic scalability} -- the combinatorial explosion of the description is avoided. However, similarly to any other rule-based language, the problem of the \emph{model execution scalability} is not covered efficiently.

The main objectives of the thesis are to develop analysis methods for rule-based models operating on the level of the model description, targeting the problem of the model execution scalability, and to develop a software tool for support of Biochemical Space Language. The main effort is to develop exact methods which provide global guarantees overcoming the limitations of approximative, sampling, and bounded methods while avoiding complete state space exploration. The emphasis is placed on model checking, parameter synthesis, and robustness analysis.

\chapter{Rule-based modelling}
\label{formal}

The conventional modelling of a biochemical system usually involves drawing a reaction-scheme diagram depicting the chemical species and reactions in the system and then manually translate it into a set of equations~\cite{voit2000computational} (e.g. ordinary differential equations). The limit of such an approach is that the models derived in this way are usually based on assumptions and simplifications. Lifting the simplifications causes a combinatorial explosion in the number of possible reactions and species~\cite{blinov2006network}, which makes manual model specification impractical.

A new starting point for the model specification is necessary. One of the potential approaches is the idea to specify the biochemical interactions as \emph{rules} that serve as generators of chemical reactions (or reaction events) and species. A rule can be viewed as a definition of a reaction class (a generalised reaction).

A rule-based system is based on \emph{rewriting} of some structured objects called \emph{agents} according to predefined \emph{rules}. The structure defined on the agents is motivated by the fact that enzymatic and binding activities of proteins involved in signalling tend to be localised to modular domains~\cite{pawson2003assembly}. Proteins may be composed of smaller parts, usually referenced as \emph{sites}, used for modification or molecular binding. Such structure enables varying the level of context given in the rules, which is generally used to restrict the domain the rule can be applied to. The less context is given, the more general the rule becomes, and it can be used in more cases. It is the key feature which supports the conciseness of the rule-based approach.

\section{Formal definition}

In this chapter, we define a simplified version of the rule-based system. The defined rules are based on simpler objects called \emph{reactions}, which are used for rewriting of unstructured objects composed by \emph{species}. Then, the rule is defined as a set of reactions, encoding multiple possibilities how it can be used, depending on a particular context. First, we start with the definition of a multiset as a form of an unstructured object used for rewriting.

\begin{definition}{Multiset}
\emph{Multiset} $\mathtt{M}$ is a pair $(\mathcal{A}, \mathtt{m})$ where $\mathcal{A}$ is a set and $ \mathtt{m} : \mathcal{A} \rightarrow \mathbb{N} $ is a function from $\mathcal{A}$ to the set of natural numbers. The set $\mathcal{A}$ is called the \emph{reference set} of elements. For each element $\mathit{a}$ in $\mathcal{A}$ the \emph{multiplicity} (that is, number of occurrences) of $\mathit{a}$ is the number $\mathtt{M}(\mathit{a})$.
\end{definition}

The elements of the multiset are often referenced as \emph{species}. We denote with $\mathbb{M}$ the set of all multisets. With $\Gamma$ we denote the set of all arbitrary rational functions over a domain of species. The notation $\nu(\mathtt{M})$ with $\nu \in \Gamma$ and $\mathtt{M} \in \mathbb{M}$ denotes the evaluation of function $\nu$ to a rational number by substituting each variable $\mathit{a}$ by its corresponding value $\mathtt{M}(\mathit{a})$.

\begin{definition}{Reaction}
A reaction $\omega$ is a triple $(\mathtt{lhs}, \mathtt{rhs}, \nu) \in \mathbb{M} \times \mathbb{M} \times \Gamma$ where $\mathtt{lhs}$ is left-hand side, $\mathtt{rhs}$ is right-hand side, and $\nu$ is a rational rate function.
\end{definition}

Informally, a reaction states how different species interact to change the number of objects of the corresponding species in the state of the system. The state of the system is an unstructured collection (multiset) of copies of each species. By $\Omega$ we denote the set of all reactions.

\begin{definition}{Reaction application}\label{react_app}
Let $\omega = (\mathtt{lhs}, \mathtt{rhs}, \nu)$ be a reaction and $\mathtt{M}$ be a multiset. The \emph{reaction application} to the multiset $\mathtt{M}$, written $\omega(\mathtt{M})$, is a pair $(\mathtt{M} - \mathtt{lhs} + \mathtt{rhs}, \nu(\mathtt{M}))$ if $\mathtt{M} - \mathtt{lhs} > 0$; $(\mathtt{M}, 0)$ otherwise.
\end{definition}

Applying a reaction to a state (multiset) means removing from the state the number of copies of elements that appear in the left-hand side ($\mathtt{lhs}$) of the reaction and adding the elements that appear on the right-hand side ($\mathtt{rhs}$) of the reaction. Additionally, the \emph{rate function} $\nu$ is evaluated in the process and provides reaction rate, i.e. speed at which the reaction happens. In the case when the number of occurrences of some species required by $\mathtt{lhs}$ is not high enough, the application is not possible and the rate is zero. For example of reaction application, see Figure~\ref{reaction_example}.

\begin{figure}[!h]
	\begin{center}
		\includegraphics[scale=0.2]{images/reaction.png}
	\end{center}
	\caption{A graphical example of reaction application. In a multiset represented by a big circle we have three kinds of agents -- {\large $\square$}, \raisebox{0.04cm}[0cm]{{$\bigcirc$}} and $\bigtriangleup$. In the reaction $\omega = (\{{\large \square}:1,\hspace{0.1cm} $\raisebox{0.04cm}[0cm]{{$\bigcirc$}}\hspace{0.1cm}$:1\}, \{\bigtriangleup:1\}, \mathtt{2}\times{\large \square}\hspace{0.05cm}\times$\hspace{0.05cm}\raisebox{0.04cm}[0cm]{{$\bigcirc$}}$)$, the agents {\large $\square$} and \raisebox{0.04cm}[0cm]{{$\bigcirc$}} are consumed and a new agent $\bigtriangleup$ is produced. The rate of the reaction is given as a function of molecular counts of both reactants. After substituting the counts to {\large $\square$} $ = \mathtt{2}$ and \raisebox{0.04cm}[0cm]{{$\bigcirc$}} $ = \mathtt{4}$, the evaluation of $\mathtt{2} \times \mathtt{2} \times \mathtt{4}$ gives the rate $\mathtt{16}$.}\label{reaction_example}
\end{figure}

\begin{definition}{Rule}
A rule $\rho \in 2^\Omega$ is defined as a set of reactions. By $\omega \in \rho$ we denote a reaction from the definition base of rule $\rho$.
\end{definition}

A rule is a generalisation of multiple reactions. It can be viewed as a high-level compact definition of a set of chemical reactions. Instead of a single possible way of application as in the case of reaction, there are multiple outcomes of the application to a given multiset. By $\Sigma$ we denote the set of all rules.

\begin{definition}{Rule application}
Let $\rho$ be a rule and $\mathtt{M}$ be a multiset. The \emph{rule application} to the multiset $\mathtt{M}$, written $\rho(\mathtt{M})$, is a set $\{ \omega(\mathtt{M}) ~|~ \omega \in \rho \}$ of all reaction applications from the definition base of the rule.
\end{definition}

Given a multiset, applying a rule to it produces a set of \emph{multiset--rate} pairs. This is caused by the fact that all possible reactions represented by the rule can produce a different outcome (we do not consider reactions with zero rate due to a non-negative number of repetitions requirement in reaction application). Example of rule application is given in Figure~\ref{rule_example}.

\begin{figure}[!h]
	\begin{center}
		\includegraphics[scale=0.2]{images/rule.png}
	\end{center}
	\caption{A graphical example of the rule application. In a multiset represented by a big circle we have three kinds of agents -- {\large $\square$}, \raisebox{0.04cm}[0cm]{{$\bigcirc$}} and $\bigtriangleup$. Let us have reactions $\omega_1 = (\{{\large \square}:1,\hspace{0.1cm} $\raisebox{0.04cm}[0cm]{{$\bigcirc$}}\hspace{0.1cm}$:1\}, \{\bigtriangleup:1\}, \mathtt{2}\times{\large \square}\hspace{0.05cm}\times$\hspace{0.05cm}\raisebox{0.04cm}[0cm]{{$\bigcirc$}}$)$ and $\omega_2 = (\{\hspace{0.1cm} $\raisebox{0.04cm}[0cm]{{$\bigcirc$}}\hspace{0.1cm}$:2\}, \{\bigtriangleup:1\}, \mathtt{3}\times$\hspace{0.05cm}\raisebox{0.04cm}[0cm]{{$\bigcirc$}}$^2)$. The rule $\rho = \{ \omega_1, \omega_2 \}$ is given by both reactions. The rule application can therefore produce a different state for each reaction, each created by particular reaction application. The associated rates are $\mathtt{2} \times \mathtt{2} \times \mathtt{4} = \mathtt{16}$ and $\mathtt{3} \times \mathtt{4}^2 = \mathtt{48}$, respectively.}\label{rule_example}
\end{figure}

\begin{definition}{Rule-based model}
A rule-based model $\mu$ is a pair $(\mathcal{R}, \mathtt{M}_0) \in 2^\Sigma \times \mathbb{M}$ where $\mathcal{R}$ is a set of rules and $\mathtt{M}_0$ is an initial state (multiset) of the model.
\end{definition}

A rule-based model is composed of two parts -- the initial state, describing the number of occurrences of individual species in the default situation, and the set of rules which define how the species can interact. The \emph{semantics} of the model is given in terms of rule application, which is transitively applied to the initial state.

A natural extension of the model is \emph{parametrised} rule-based model where rate functions can be enriched by unknown parameters from a known domain (usually integers or rational numbers). It is a typical case in systems biology since the actual values of reaction kinetics are often hard or even impossible to be measured. Then, to evaluate a rate function, a \emph{parametrisation} $p$ from the admissible parameter space has to be given, representing a particular biological scenario.

\section{Analysis methods}

An important advantage of the rule-based approach is that mathematical models can be automatically generated from it. In particular, instead of relying on a single mathematical formalism, different mathematical models can thus be obtained for a given model (e.g., ODEs~\cite{camporesi2017k}, chemical master equation or continuous-time Markov chains~\cite{pauleve2010generic,sneddon2011efficient}). This fact results in multiple possible ways of how to analyse and explore the models. 

\emph{Simulation} is the most usual way how to exploit the behaviour of a model. Depending on the interpretation of rule rates, we can use numerical simulation of the rule-based model for both its continuous and stochastic dynamics. Analysing the models using simulation is very useful, but does not provide a global overview of its behaviour.

\emph{Model checking}~\cite{clarke2018model} is an automatic way for checking whether the model meets a given specification. For that, we need a specification language which allows us to express a property dependent on time. A suitable tool is to use a \emph{temporal logic}. Examples of logics which are often used in systems biology are LTL~\cite{pnueli1977temporal}, CTL~\cite{clarke1986automatic}, STL~\cite{maler2004monitoring}, or PCTL~\cite{hasson1994logic} and their usage depends on particular mathematical representation we use.

\begin{definition}{Model checking}
Given a rule-based model $\mu = (\mathcal{R}, \mathtt{M}_0)$ and a property formula $\phi$, the \emph{problem of model checking} is to decide whether the model $\mu$ satisfies the property $\phi$.
\end{definition}

In the case of unknown parameter values present in the model, we can be interested in finding those parametrisations from the parameter space which satisfy the given property. For that, a method called \emph{parameter synthesis} is used. The basic principle is to search parameter space and identify regions where satisfaction is (not) guaranteed. Often regions, where no guarantees are given, are produced due to some computational error of used searching method.

\begin{definition}{Parameter synthesis}
Given a parametrised rule-based model $\mu = (\mathcal{R}, \mathtt{M}_0)$ and a formula $\phi$, the \emph{problem of parameter synthesis} is to compute a partitioning of the given parameter space into three disjoint subsets: $\mathsf{TRUE}$ -- the model satisfies the property, $\mathsf{FALSE}$ -- the model does \emph{not} satisfy the property, and $\mathsf{UNKNOWN}$ -- the result is not known.
\end{definition}

Additionally, instead of the binary answer to the satisfaction of a property, a measure of how much the property is preserved in the given parametrisation can be useful. This measure is called \emph{local robustness} $D^{\mu}_{\phi}(p)$ and states how much the property $\phi$ is preserved in parametrisation $p$ of model $\mu$.

\begin{definition}{Robustness analysis}\label{robustness_def}
The \emph{problem of global robustness} of a model $\mu$ is defined as 

$$ R^{\mu}_{\phi, P} = \int_{P} \psi(p) D^{\mu}_{\phi}(p) dp $$

where $\phi$ is the property of the system under scrutiny, $P$ is the parameter space, $\psi(p)$ is the probability of the parametrisation $p$, and $D^{\mu}_{\phi}(p)$ is the local robustness.
\end{definition}

The \emph{global robustness}~\cite{kitano2007towards} defines a measure of how the given property is globally preserved in the model with respect to the parameter space.

\chapter{State of the Art} \label{chap:state}

In this chapter, we provide an insight into the state-of-the-art of the rule-based modelling, including its foundations, existing available formalisms, and established analysis techniques which can help in the process of model-building tasks.

The formal foundation of rule-based modelling is based in process algebra that enables principled mathematical reasoning about system behaviour and its dependency on interaction capabilities of system components. One of the first used approaches is $\pi$-calculus~\cite{milner1999communicating}, a general and minimalist model-specification language originally designed to capture essential features of concurrent and distributed systems in computer science. Use of $\pi$-calculus to model biological interactions was suggested in~\cite{regev2000representation}.

However, because $\pi$-calculus is a minimalist language, and some of its notational features are irrelevant and even inappropriate for biological applications (for example, communication has directionality, whereas physical association does not), researchers have sought to develop a more congruent higher-level language for
modelling biochemical systems that retains the mathematical formality of $\pi$. In the next section, we provide a list of the most established rule-based formalisms.

\section{Rule-based languages}
\label{rule_based_languages}

There are several rule-based languages dedicated to modelling of the biological systems. Each of them uses different features and abstractions. In this chapter, we will highlight the key features of some of the representatives and describe analysis established for them. It is important to note that generally, they can always be formulated in terms defined in Chapter~\ref{formal}.

A specificity of individual languages resides in the level of abstraction they use and the way how they handle the encoding of a rule such that they avoid the explicit reaction network representation. The implicit representation they use includes both syntactic and semantic level, providing a direct apparatus of rules execution.

\subsection{Kappa language}
\label{kappa}

The Kappa language~\cite{kappa_formal} was primarily developed for the modelling of protein interactions. The key structures used in the language are \emph{agents} with \emph{binding sites}, which allow formation of \emph{bonds} between the agents. Each binding site must be unique with at most one bond. Each site can occur in one of several pre-defined states.

The Kappa \emph{rules} are changing properties of the agents. Particularly, the rule might add, delete, or change a bond or a state of one or multiple agents at once. The rules are patterns with two sides where each of them is a sequence of agents delimited by a comma. The introduction of sites allows to group multiple reactions to a single rule by omitting the context which is not relevant. This is a typical way how rule-based languages encode reactions to rules and it ensures conciseness of description. Example of a rule is in Figure~\ref{kappa-rule}.

\begin{figure}[!h]
\begin{center}
\begin{tabular}{c l}
(a) & A(bsa$\sim$\{ u, p \}) ; B(bsb$\sim$\{ a, i \}) \\
(b) & A(bsa), B(bsb$\sim$a) $\rightarrow$ A(bsa!1), B(bsb$\sim$a!1) \\
\end{tabular}
\end{center}
\caption{An example of a Kappa rule. (a) Definition of associated agents. There are allowed two agents, A and B, both with a binding site which might occur in two possible states. (b) A Kappa rule which defines a creation of bond on the sites of agents A and B such that site of agent B must be in an active state. Note the context with no importance for the interaction is omitted, encoding multiple possible reactions.}\label{kappa-rule}
\end{figure}

The general problem present in the Kappa language is too detailed description. For biological systems, it might often be difficult to manage bonds between sites, especially in cases when we do not need such details.

\subsection{BioNetGen Language}
\label{bngl}

The BioNetGen Language (BNGL) \cite{harris2016bionetgen} is similar to the Kappa language, but there are several differences. The basic idea of the language is to use molecules to describe the building blocks of a biological system such as proteins, genes, and metabolites. Each molecule has a number of sites with associated internal states that are used to represent the status of post-translational modifications or bindings with other molecules. The rules describe the interactions among molecules including associaÂ­tions, dissociations, modifications to the internal state of a site as well as the production or consumption of molecular species. Rules also provide rate laws for transformations resulting from molecular interactions. Patterns are used to identify a set of molecules that share the same internal states. The rule can be specified using patterns such that it defines not a single reaction but a potentially large class of reactions, all involving a common transformation parametrised by the same rate law.

BNGL rule typically represent molecular interactions and the consequences of these interactions. Example of a rule is given in Figure~\ref{bngl-rule}. Software tools have been developed to construct, visualise, simulate, and analyse BNGL models~\cite{wenskovitch2014mosbie,harris2016bionetgen,xu2011rulebender,sneddon2011efficient}.

\begin{figure}[!h]
\begin{center}
\begin{tabular}{c l}
(a) & A(bsa$\sim$u$\sim$p) \\
  & B(bsb$\sim$a$\sim$i) \\
(b) & A(bsa) + B(bsb$\sim$a) $\rightarrow$ A(bsa!1).B(bsb$\sim$a!1) \\
\end{tabular}
\end{center}
\caption{An example of a BNGL rule. (a) Definition of associated agents. There are allowed two agents, A and B, both with a binding site which might occur in two possible states. (b) The rule itself. Note that the agents A and B must first create \emph{reaction complex} denoted by `+' (i.e., they must be close to each other) and then they create \emph{complex of agents} denoted by `.' (i.e., they are physically connected).}\label{bngl-rule}
\end{figure}

\subsection{PySB}

The PySB language~\cite{lopez2013programming} is an open-source programming framework written in Python that allows concepts and methodologies from contemporary software engineering to be applied to the construction of transparent, extensible and reusable biological models. It is implemented as a package in the Python programming language. 

The definition of the models directly in the code allows the usage of the full syntax of Python, which significantly increases the expression power of PySB. Its widespread use in the computational biology community, support for object-oriented and functional programming, and rich ecosystem of mathematical and scientific libraries make it an excellent choice of programming language for this purpose. On the other hand, the increased expressive power make the models harder to analyse.

The core of the language is defined by translating to BNGL. PySB is closely integrated with Python numerical tools for simulation and parameter estimation and graphical tools that enable plotting of model trajectories and topologies. The main advantage of the language is that it can be used to decompose models into reusable macros that can be independently tested, and the used to generate composite models. Example of a rule is given in Figure~\ref{pysb_rule}.

\begin{figure}[!h]
\begin{center}
\begin{python}
# Declare the monomers
Monomer('L', ['s'])
Monomer('R', ['s'])

# Declare the binding rule
Rule(L(s=None) +  R(s=None) <> L(s=1) % R(s=1), kf, kr)
\end{python}
\end{center}
\caption{An example of a PySB rule. The rule describes creation of a bond between agents $\mathtt{R}$ and $\mathtt{L}$.}\label{PySB-rule}\label{pysb_rule}
\end{figure}

\subsection{Chromar}

The Chromar language~\cite{honorato2018chromar} allows to define attributes for agents and range them over pre-defined domains. The qualitative semantics is given by rule match on multisets composed of these agents producing a reaction. It is followed by standard application of the reaction (in the manner of multiset operations, for details see Definition~\ref{react_app}). The language is very useful when creating a model where we need to create new distinct objects and control the population of these objects.

Embedding this language into the functional programming language Haskell increases its expressive power while making the ability of some analysis more expensive. Moreover, a user needs to understand Haskell (at least its basics) in order to use Chromar.

It is important to highlight a feature which offers the stochastic semantics of this language. Compared to the other rule-based languages, it is capable of specifying the rates for individual reactions inherited from the rule (Figure~\ref{chromar_rule}). It is allowed by variable value bindings and type-determination between the left and right-hand side of the rule.

\begin{figure}[!h]
\begin{center}
A(a = x), A(a = y) $\xrightarrow[]{\text{f(x,y)}}$ A(a = x + y), A(a = y - 1) [g(x, y)]
\end{center}
\caption{An example of a Chromar rule. Arithmetic operations can be used in order to change the properties of individual agents. Moreover, a rate function f and conditional function g increase applicability and practical use of the rule.}\label{chromar-rule}\label{chromar_rule}
\end{figure}

However, when it comes to readability and presentation to the user, the language has a crucial disadvantage -- all the biologically relevant terms such as states have to be encoded in natural numbers.

\subsection{Language for Biochemical Systems}

Language for Biochemical Systems (LBS)~\cite{Pedersen} combines rule-based approach-es to modelling with modularity. The main features of the language are species expressions for manipulating large complexes in a concise manner, parameterised modules with a notion of subtyping for writing reusable modules, and nondeterminism for handling combinatorial explosion.

A simple example of an LBS program is shown in Figure~\ref{LBS_example}. Species identifiers such as \emph{mrna} are here assigned to new species. The text \textbf{new{}} is formally a species expression which evaluates to a species value with no modification sites and with a globally unique name. Hence the name of the species identifier, \emph{mrna}, does not in itself hold any identity of a species, and one may bind the same identifier to an entirely different species in another part of the program. This allows different modules, possibly developed by different people, to use the same species identifier for mRNA molecules which are semantically and biologically different, and subsequently combine the modules into a single program without unintended cross-talk.

\begin{figure}[!h]
\lstset{language=LBS}
\begin{lstlisting}[basicstyle=\scriptsize, frame=single, escapeinside={(*}{*)}]
spec gene = new{}, rnap = new{}, mrna = new{}, prot = new{}; 
comp c new comp; comp n = new comp inside c;
c [ 
	n [ gene + rnap (*$\rightarrow$*) gene + rnap + mrna ] | 
	n  [ mrna ] (*$\rightarrow$*) mrna | 
	mrna (*$\rightarrow$*) prot
]
\end{lstlisting}
\caption{An LBS program for gene expression.}\label{LBS_example}
\end{figure}

In~\cite{Pedersen}, authors provide a demonstration of multiple concrete semantics defined for the rule-based description; namely basic Petri nets, coloured Petri nets, ordinary differential equations, and Continuous-time Markov chains. For that, they usually use a reaction-based (or ground) representation of the system (except coloured Petri nets which allow most of the rule-based abstract features to be encoded in colours).

Moreover, LBS-Kappa~\cite{pedersen2015high} is a modular extension of Kappa language based on LBS, providing a language for writing high-level and modular rule-based models. Since LBS is a general framework, it was only needed to define the necessary instantiation to Kappa and refer to the general semantics of LBS. 

\subsection{Biochemical Space Language}

Biochemical Space Language (BCSL)~\cite{trojak2018sasb} is a language that combines the advantages of different approaches and thus makes an effort to overcome several problems of existing solutions. BCSL is an integral part of a general framework developed for comprehensive modelling in systems biology~\cite{klement2013comprehensive,BCS}. BCSL relies on the formal basis of the rule-based methodology while preserving user-friendly syntax of plain chemical equations. The target group of users is a biologically-oriented community. Therefore the emphasis is placed on usability and positive user experience.

The language uses several types of agents, describing different levels of abstraction and detail. Particularly, there is an \emph{atomic} agent describing the most basic type of biological objects which have assigned an internal state. The state typically represents a modification of the objects, e.g. oxidised, reduced, methylated etc. \emph{Structure} agent represents a biochemical object that is composed of several known atomic agents while we know that composition is abstract and not necessarily complete. Since not all biochemical structures are known in detail, this feature is useful in expressing partial knowledge (for example, see Figure~\ref{ps2_rule}). \emph{Complex} agent groups several structure or atomic agents, forming a non-trivial composite biochemical object. A typical feature of the language is that it does not use any form of binding when it comes to complex formation. Instead, the complex formation is considered as a form of \emph{coexistence}; the particular interpretation depends on the context of the model.

\begin{figure}[!h]
  \begin{center}
    $ \mathit{ps2}(oec\{$\emph{2+}$\}, yz\{$+$\}){::}tlm \Rightarrow \mathit{ps2}(oec\{$\emph{3+}$\}, yz\{$\emph{n}$\}){::}tlm $
  \end{center}
  \caption{An example of BCSL rule. Oxidation of S2-state of the oxygen-evolving complex $oec\{$\emph{2+}$\}$ by $Yz\{+\}$ in photosystem II ($\mathit{ps2}$). The photosystem is represented as a structure agent. There are many other subparts of photosystem II. (e.g. chlorophyl, pheophytin), but their states are not important in this case. This makes the syntax of the language very concise and readable.}\label{ps2_rule}
\end{figure}

The agents interact via rules. In the definition of a rule, identifiers of \emph{substrates} and \emph{products} are used to make the notation of the rules compact. Every agent appearing in a rule equation has to be followed by the \emph{localisation} operator associating it with a particular compartment. This is, for example, important for rules that act on both sides of a membrane. That way, a rule is always precisely localised in or between the compartments.

\begin{figure}[!h]
	\begin{center}
		$ S\{u\}{::}KaiC{::}?X{::}cyt \Leftrightarrow S\{p\}{::}KaiC{::}?X{::}cyt ~;~$\vspace*{0.3cm}
		
		$ ?X = \Set{
			\begin{array}{c}
			\mathit{KaiC6}, \mathit{KaiA2C6}, \mathit{KaiA4C6},\\
			\mathit{KaiB6C6}, \mathit{KaiA4B6C6}, \mathit{KaiA6B6C6}
			\end{array}
		}$
	\end{center}
	\caption{An example of BCSL rule. The state of a serine residue $S$ of $KaiC$ protein can be changed from unphosphorylated to phosphorylated whenever the $KaiC$ protein is included in a complex. The particular allowed complexes which enable this modification are given by the variable $?X$. Each of them is provided in the form of a complex alias, which requires a proper complex definition in the model.}
	\label{BCSL_rule}
\end{figure}

Moreover, it is possible to assign a name to complex, which increases the readability and provides additional annotation options. This is supported by the binary localisation operator `{::}'. The main idea is to allow zooming into individual parts of agents. The localisation operator naturally leads to an extension by \emph{variables}, which can group several rules matching in the same modification of an agent (for example, see Figure~\ref{BCSL_rule}).

\subsection{SBML-multi package}

Systems Biology Markup Language (SBML)~\cite{hucka2003systems} is a standard established for systems biology based on XML. It is popular standardised format for the electronic storage, exchange, and reuse of mathematical models of biochemical systems. However, it is based on the assumption that a model can be specified adequately in terms of a reaction scheme, which is not suitable for a rule-based model.

The newest version SBML Level 3 introduced modular language extension capability, which allows different language features to be added to a common language core. An SBML-multi package~\cite{SBMLmulti} is able to describe all the necessary rule-based features, and therefore, it is possible to export each model in a rule-based language in this format. It is the most suitable format for the exchange and storage of the models but less for analysis and direct representation to the users. It serves as an intermediate format for model sharing. The presence of a feature tag informs a software tool reading the model that the model uses that particular feature and permits the tool to quit if it does not have the necessary interpretive capabilities. For example, see Figure~\ref{SBML_example}.

\begin{figure}
\lstset{language=XML}
\begin{lstlisting}[basicstyle=\scriptsize, frame=single]
<?xml version="1.0" encoding="UTF-8"?>
<sbml xmlns="http://www.sbml.org/sbml/level3" level="3" version="1">
  <model>
    <listOfCompartments>
      <compartment id="c" constant="true" multi:isType="true" />
      <compartment id="cc" constant="true" multi:isType="true">
      ...
        <multi:listOfCompartmentReferences>
          <multi:compartmentReference multi:id="cr1" multi:compartment="c" />
          <multi:compartmentReference multi:id="cr2" multi:compartment="c" />
          ...
        </multi:listOfCompartmentReferences>
      </compartment>
    </listOfCompartments>
    <multi:listOfSpeciesTypes>
      <multi:bindingSiteSpeciesType multi:id="stA" multi:compartment="c" />
      <multi:speciesType multi:id="stAA" multi:compartment="cc">
        <multi:listOfSpeciesTypeInstances>
          <multi:speciesTypeInstance multi:id="stiA1" multi:speciesType="stA"
            multi:compartmentReference="cr1" />
          <multi:speciesTypeInstance multi:id="stiA2" multi:speciesType="stA"
            multi:compartmentReference="cr2" />
          ...
        </multi:listOfSpeciesTypeInstances>
        <multi:listOfInSpeciesTypeBonds>
          <multi:inSpeciesTypeBond multi:bindingSite1="stiA1" multi:bindingSite2="stiA2"/>
          ...
        </multi:listOfInSpeciesTypeBonds>
      </multi:speciesType>
    </multi:listOfSpeciesTypes>
    <listOfSpecies>
      <species id="spA" multi:speciesType="stA" compartment="c" ... />
      <species id="spAA" multi:speciesType="stAA" compartment="cc" ... />
      ...
    </listOfSpecies>
    <listOfReactions>
      <reaction id="reaction" ...>
        <listOfReactants>
          <speciesReference id="r1" species="spA" multi:compartmentReference="cr1" ... />
          <speciesReference id="r2" species="spA" multi:compartmentReference="cr2" ... />
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="spAA" ... />
        </listOfProducts>
        ...
      </reaction>
      ...
    </listOfReactions>
  </model>
</sbml>
\end{lstlisting}
\caption{An example of a simple SBML-multi model. The model requires specification of language level (version). In the definition, there are included SpeciesTypes, Species which belong to these types, and Reactions where those species are interacting. Additionally, all processes are encapsulated in the compartments.}\label{SBML_example}
\end{figure}

\section{Analysis methods}\label{analysis_sec}

In the context of systems biology, the computer-aided verification methods are becoming important for analysing the models, validating new experimental results, automatically checking behaviours of interest, and identifying the inputs or parameters of the system enforcing the desired behaviour. The formal verification of a model consists of proving that its execution satisfies a given specification of the required behaviour.

Rule-based models can be used to generate other computational models using different semantics. Therefore, any developed analysis techniques for the other computational models can be indirectly used to analyse the rule-based models. A disadvantage of such a procedure is that the computational models are often explicit, and thus the implicit representation of the rule-based approach is lost during the translation. 

\subsection{Simulation}

The most usual way how to exploit the behaviour of a rule-based model is a simulation. While qualitative simulation is also possible (generate a possible successor of the current state abstracting while the notion of time), more usual is a quantitative simulation. Using the generated reaction network from the model rules, one can perform stochastic simulation, for example, using Gillespie algorithm~\cite{GILLESPIE1976403}) or deterministic simulation~\cite{Poole2000} by consequent translation to Ordinary differential equations (ODEs)~\cite{higham2008modeling}. However, the cost of simulating the kinetics of a reaction network depends on the size of the network. Thus, if a network is large, the simulation cost can be expensive in terms of computation time and/or computer memory. To address this problem, network-free simulators were developed that take advantage of rules. The method generalises an agent-based kinetic Monte Carlo method that has been shown to circumvent the combinatorial bottleneck in simulations~\cite{yang2008kinetic}.

An example of a network-free simulator is NFsim~\cite{sneddon2011efficient}. The advantage of NFsim compared to reaction-based approaches is that during simulation, rules operate directly on molecular objects to produce exact stochastic results with the performance that scales independently of the reaction network size. Reaction rates can be defined as arbitrary functions of molecular states to provide powerful coarse-graining capabilities. NFsim enables researchers to simulate many biological systems that were previously inaccessible to general-purpose software.

Analysing the models using simulation is very useful, but usually not sufficient. The models are often difficult to calibrate due to many unknown parameters and limited experimental training data. It is necessary to apply the powerful tools developed in the context of program verification to biological models, particularly when
they can be cast in the form of rules that are executed stochastically on the fly. The simulation will only yield a particular trajectory at each run. Even when many runs are gathered to perform statistical analysis, observing a time series of concentrations (or molecule numbers) does not necessarily lead to an understanding of the model.

\subsection{Model checking}

A typical approach of dynamic analysis is to compute a data structure, describing the whole system behaviour. The easiest way to do this is by constructing the \emph{reachability graph} (also called state space). The nodes of a reachability graph represent all possible states of the modelled system and the arcs represent actions which caused the particular state change. However, reachability graphs tend to be extremely large because they comprise all possible system states. The state space explosion motivates the static analyses discussed below in this section. 

If we succeed in constructing the complete reachability graph, we can decide the behavioural properties of the modelled system. Here are some typical examples of qualitative dynamic analysis:

\begin{itemize}
	\item \emph{Reachability}. The reachability problem is to check whether there exists a path from a given state to a particular state. The problem can be reformulated to finding a sub-state, meaning we are only interested in the molecular counts of particular species. It has been shown that the reachability problem is decidable~\cite{kosaraju1982decidability} although it takes at least exponential space (and time) to verify in the general case~\cite{lipton1976reachability}.
	\item \emph{Boundedness}. A system is bounded if the number of molecular counts is finite for any state reachable from the given initial state. This is very useful for the detection of undesirable inconsistency in the flows in the model. It is also generally required by model checking.
	\item \emph{Persistence}. A system is persistent if, for any two reaction events, the firing of one reaction will not disable the other. This analysis is suitable for detecting concurrent reactions in the models, which are often responsible for some critical decision-making in biological systems.
\end{itemize}

Generally, checking a property of a model is called \emph{model checking}. Model checking is an automatic formal verification technique able to perform a clever exhaustive search of the state space of a model. It operates over a discrete-time model with a finite number of states (usually a Kripke structure~\cite{kripke1963semantical}), determining the truth value of a temporal logic formula which specifies desired property of the modelled system. Examples of logics which are often used in systems biology are LTL~\cite{pnueli1977temporal}, CTL~\cite{clarke1986automatic}, STL~\cite{maler2004monitoring}, or PCTL~\cite{hasson1994logic}, depending on the particular mathematical formalism.

A rule-based model can be translated to such a formalism and analysed by an efficient model checker such as~\cite{cimatti2002nusmv}. An important issue with this technique is that the number of states of a model usually grows exponentially in the number of its parameters, giving rise to the state explosion problem. Additionally, in the general case, a state space of the rule-based is infinite, making the analysis even harder to handle. For this reason, there are often some limitations in the application of the rules which bound possible state space (e.g. global bound on allowed molecular counts in states), abstracting some information and providing a way how to create a finite state space.

For these reasons, most of the model checkers usually represent the states symbolically~\cite{burch1992symbolic} such as binary decision diagrams~\cite{bryant2001graph}. Relevant methods for the analysis of continuous semantics are based on the abstraction of the behaviour to a finite number of discrete states. For example, one can be interested in parallel semi-symbolic coloured model checking~\cite{benevs2016model}, bifurcation analysis~\cite{benevs2017discrete}, and attractor analysis~\cite{benevs2018fully}.

Model checking has been extended to many other computational models such as continuous- and discrete-time Markov chains (CTMC and DTMC) by adding probabilities~\cite{kwiatkowska2010probabilistic}. A tool used for analysing the Markov chains is PRISM~\cite{kwiatkowska2011prism}, which calculates the probability that the model will satisfy the property of interest. It can either provide the exact solution~\cite{aziz2000model} or an approximated solution~\cite{jha2009bayesian} (using a set of samples, generated using a Monte Carlo simulation of the model). Moreover, it is also possible to compute a numerical value expressing the measure of satisfaction instead of a binary decision, often referred to as \emph{quantitative model checking}~\cite{huth1997quantitative}.

\emph{On-the-fly} algorithms employ a top-down approach to model checking, which does not require global knowledge of the complete state space~\cite{latella2014fly,bhat1995efficient,courcoubetis1992memory,gnesi2011abstract,holzmann2004spin}. Instead, the algorithms construct step by step local knowledge of the state space until it is possible to decide whether the given state satisfies the formula. 

\subsection{Parameter synthesis}\label{synthesis}

\emph{Parameter synthesis} is the inverse problem to model checking -- its goal is to find the maximal subset of parameter values such that they meet the stated dynamical constraints. In practice, the unknown parameters are usually limited by a set of admissible values, providing a parameter space. In the context of systems biology, there are several relevant methods which solve parameter synthesis. 

In~\cite{barnat2011parameter}, the authors present an algorithm for parameter synthesis based on parallel model checking, which is conceptually universal with respect to the modelling approach employed. In~\cite{batt2010efficient}, an efficient method to analyse large and possibly incomplete parametrised piecewise-affine differential equation models is proposed. Usage of a symbolic encoding of the model structure enables the methods to take full advantage of symbolic model checkers for testing CTL dynamic properties. An algorithm for computing parameter synthesis in non-linear dynamical systems~\cite{donze2010parameter} extends formal verification techniques that were first introduced in the context of continuous and hybrid non-linear dynamical systems~\cite{donze2007systematic}.

In~\cite{jha2011synthesis}, algorithms for studying the parameter space of stochastic biochemical models were presented. Methods for parameter synthesis of parametric Markov chains have been introduced with symbolic computation of reachability properties through state elimination~\cite{daws2004symbolic,jansen2014accelerating,hahn2011probabilistic}, recently improved by parameter lifting~\cite{quatmann2016parameter} and fraction-free Gaussian elimination~\cite{hutschenreiter2017parametric}.

An alternative approach to the analysis of complex stochastic models under
parameter uncertainty is based on statistical methods~\cite{Luck2016,BackenkohlerBW18,Bortolussi2016,Bortolussi2018}. There are only a few works that bridge the rule-based framework to such techniques. In~\cite{liu2016parameter}, a statistical parameter sampling method is employed to analyse unknown parameters in BNGL models represented by means of CTMCs where the rate function is limited to mass action kinetics. The recent work~\cite{KJ18} combines statistical model checking with machine learning techniques to calibration (estimation) of parameters in order to maximise the probability of satisfying a given specification. In~\cite{Bock15} the authors adapt simulation-based and moment-based methods. In general, statistical techniques do not give an exact symbolic representation of satisfying parameter sets. 

\subsection{Monitoring}

Another way to tackle the state explosion problem is the analysis of a single execution trace instead of performing an exhaustive verification. \emph{Monitoring} aims to check whether the current execution of a program satisfies or violates a property of interest. For example, in BIOCHAM~\cite{calzone2006biocham}, authors implemented monitoring of numerical simulations of biological models with LTL used to specify properties of real-valued variables. Additionally, in~\cite{donze2010breach} they implemented a tool where the evaluation of the STL~\cite{maler2004monitoring} formula robustness is used for a particular trajectory through monitoring.

\subsection{Robustness analysis}

According to~\cite{kitano2004biological}, robustness is a property that allows a system to maintain its functions against internal and external perturbations. The formal definition of robustness is given in Definition~\ref{robustness_def}. The concept of robustness is well established for deterministic systems~\cite{donze2011robustness,rizk2009general}. The evolution of a stochastic system is given
by a set of paths compared to a single trajectory of a deterministic system. The stochastic system at any given time is described by a probability distribution over states of the corresponding CTMC in contrast to the single state representation of a deterministic system. Therefore, the definition of robustness for stochastic systems requires a more sophisticated approach. In~\cite{vceska2014robustness}, authors developed an adaptation of the concept of robustness to stochastic systems, particularly robustness analysis in CTMCs. The method is based on a numerical approximation of the evaluation function~\cite{brim2013exploring}, which provides accuracy guarantees in contrast to existing methods employing parameter sampling and statistical techniques.

\subsection{Static analysis}\label{static_analysis}

Static analysis is performed on the description of the model without the need to actually execute it~\cite{nielson2015principles}. While model checking generally needs to explore all the states originated by executing the semantics of the model, static analysis operates on the syntactic level of the specification or by using abstract interpretation~\cite{cousot1977abstract} over finite approximations of the possible model executions~\cite{pauleve2012static}. It is a powerful tool preferably used to detect potential issues in the model.

There are several well-defined static analysis methods for rule-based systems, which can be useful in finding inconsistency in the models~\cite{danos2009rule}. For example: \emph{detection of dead rules} -- a rule is called dead, if there is no trace starting from the initial state in which this rule is applied; \emph{detection of dead agents} -- an agent is called dead, if there is no trace starting from the initial state with at least one state in which this agent occurs; \emph{detection of redundant rules} -- a rule is redundant if after removing it from the model, the particular semantics do not change. These and many more methods can be run statically, providing an efficient way of how to verify special types of properties before executing the model.

To the best of our knowledge, the only tool dedicated to static analysis of rule-based models is KaSa~\cite{boutillier2018kasa}. This static analyser abstracts the set of reachable states of models and then uses this information to collect insightful properties. In particular, KaSa may warn about rules that may never be applied, about potential definitive transformations of proteins, and about the potential formation of unbounded molecular compounds.

\chapter{Aim of the Thesis} \label{chap:aim}

In general, there exist three approaches to analysis of rule-based models: (i) \emph{direct} approach operating on the level of the model specification, (ii) \emph{indirect} approach analysing the transition system generated from the model, and (iii) \emph{semi-direct} approach employing heuristics or a symbolic representation in the process of analysis.

One of the drawbacks of existing analysis methods of rule-based systems is the lack of direct approaches. There are multiple ways how to analyse them indirectly through existing computational methods established in computer science (Section~\ref{analysis_sec}). However, most of them require a transition system of the modelled system, which opposes the scalability of the rule-based approach. Direct methods, like those provided by static analysis (Section~\ref{static_analysis}) or some other symbolic methods, utilise the abstractions of the approach, avoiding explicit representation of the transition system. 

The main objectives of the thesis are to develop analysis methods for rule-based models using either direct or semi-direct approach and to develop a software tool for support of Biochemical Space Language, as a representative of rule-based languages suitable for modelling in systems biology.

\section{Objective I: Analysis methods}

The first objective of the thesis is to propose direct or semi-direct analysis methods for rule-based models, which should be general and applicable to any rule-based formalism. The focus is put on model checking, parameter synthesis, and robustness analysis.

\subsection{Model checking}

To the best of our knowledge, the existing model checking methods for rule-based languages use either the indirect or the semi-direct approach. Since there do not exist direct methods for model checking of rule-based models, the primary goal is to establish a direct method. Alternatively, an improvement of existing semi-direct approaches such that they provide exact results (no heuristics) is also possible. The emphasis is placed on the reachability problem, as one of the most important problems when it comes to the analysis of biological systems.

Computer science already offers several approaches for model checking of models, which can be potentially used in the case of rule-based models. One option is to use model checking methods developed for reaction-based systems which avoid the exploitation of entire state space. It is necessary to adapt these methods for rule-based formalism.

Another option would be to use an existing indirect approach which requires complete state space and encode it symbolically or to explore on-the-fly techniques. For example, there is a method for probabilistic reachability analysis of DTMC, which constructs a regular expression (RE) using state elimination. The evaluation of RE gives the probability of reaching targeted species. The algorithm for RE construction, however, does not necessarily need the explicit DTMC. We could use this fact for an on-the-fly construction of the RE, potentially avoiding the complete state space enumeration.

The ideal solution avoids the building of the state space completely. It is the most challenging task and requires the development of new symbolic and static methods. However, as it was shown in~\cite{trojak2018sasb}, some of the model checking subtasks can be solved statically due to unique abstraction employed by rule-based languages (e.g., some cases of non-reachability).

\subsection{Parameter synthesis}

Parameter synthesis is a crucial branch of model analysis in the case when the parameters of the modelled system are not completely known. However, in rule-based community, the direct approach to this problem is not yet well developed. The only known work uses the semi-direct approach and focuses on parameter estimation using statistical model checking (Section~\ref{synthesis}). Therefore another goal of this thesis is to develop parameter synthesis algorithm for rule-based models which provides exact results avoiding the indirect approach.

One of the options on how to solve this problem is based on model checking approach from the previous subsection. Once we have an efficient model checking method, the parameter synthesis can be formulated using this method. For example, the mentioned method which constructs RE could be generalised to the parametrised case. Instead of a single probability evaluation, we obtain a probability function of parameters, which can be then used for parameter space exploitation. 

Other solutions could, for example, use some heuristic to obtain probability function, avoiding the usage of model checking, but providing only limited guarantees on the correctness of the results.

\subsection{Robustness analysis}

Robustness analysis characterises the mean validity of a formula over all parameter values in the given perturbation set. To the best of our knowledge, this problem was not yet formulated for rule-based systems. Another goal of this thesis is to define and solve this problem in the context of rule-based systems. 

In order to provide an exact solution, it is necessary to compute the function of parameters. One of the possible solutions is using RE approach mentioned in previous sections. Having such a function, we can integrate over the perturbation set and compute the robustness. 

Another possible way how to achieve this goal is by using heuristics to estimate the local robustness, which would provide an approximate estimation of global robustness.

\section{Objective II: Software tool}

The second objective of the thesis is to develop tool BCSgen to support modelling and analysis of Biochemical Space Language, which includes the implementation of the proposed analysis methods. In order to enable the proposed analysis of the models written in the language, an extension of the language by quantitative properties is necessary, which are not supported in the present version yet.

BCSgen implements environment suitable for editing and maintenance of the BCSL models by providing the following functionality: an editor with syntax highlighting and error messaging in the process of model building; simulation engine offering both deterministic and stochastic simulations; reaction network generator; transition system generator; export and import of models in appropriate standard file format; visualisation of transition system including highlighting paths from the initial state to other states, which satisfy reachability condition; visualisation of time-series data from simulation; all the proposed techniques for model checking, parameter synthesis, and robustness analysis.

The tool will be integrated within CMP, improving the user's model building experience. It is supported by the fact that BCSL is targeted to be so-called \emph{human-readable} and simple enough such that it is understandable beyond the computer science community. The languages from Section~\ref{rule_based_languages} all offer interesting features when it comes to modelling. However, none of them except BCSL focuses on these properties.

\newpage
\section{Progression schedule}

Plan for future study and research activities:

\begin{itemize}
\item Fall 2019: Doctoral exam and defence of this thesis proposal.
\item Fall 2019 -- Fall 2020: Development of analysis methods.
\item Fall 2020 -- Spring 2021: Implementation of developed methods.
\item Spring 2021: Release of a stable version of the software tool.
\item Spring 2021 -- Fall 2021: Work on the text of the PhD thesis.
\item Fall 2021: Thesis defence.
\end{itemize}

% ------------------------------- bibliography

\chapter{Achieved results}

Our research has been primarily focusing on the development of new modelling approaches in systems biology. Our long-term ongoing research is focused on the development of Comprehensive Modelling Platform (CMP) -- a general framework for public sharing, annotation, and visualisation of domain-specific biological models. For a selected system (e.g. an organism, particular process), the framework is instantiated as a web-based application which allows capturing several aspects of biological models. The platform is targeting biologically-oriented users and is trying to improve model-building tasks, including analysis support and automatic verification of models.

In~\cite{trojak2016cyanobacterium}, we presented e-cyanobacterium.org\footnote{\href{https://www.e-cyanobacterium.org}{https://www.e-cyanobacterium.org}} -- an instance of CMP oriented on processes of cyanobacteria. The platform is unique in focusing purely on cyanobacteria, which allows integration of the knowledge and its presentation in a concise and understandable form. Several well-annotated and curated models of particular cyanobacteria processes developed by the consortium are provided within e-cyanobacterium, covering environmental processes, respiration and photosynthesis, carbon concentrating mechanism, circadian clock, and metabolism.

We used the models from e-cyanobacterium to automatically analyse attractors~\cite{benevs2018fully}, which can characterise complex dynamics arising in biological systems. We demonstrated a novel method for detecting terminal strongly connected components in the ODE models, which can be abstracted in the form of parametrised graphs. The main advantage of the method is the ability to detect attractors without precise knowledge of systems behaviour or even its parameters. This type of analysis can be performed to discover unstable or oscillatory behaviour in a model where such behaviour is typically undesired. In Figure~\ref{attractors}, there are some interesting results from this case study. 

\begin{figure}[!h]
\begin{minipage}{0.45\textwidth}
\begin{center}
\includegraphics[scale=0.3]{images/clark_fast_vs_co2.png}
\end{center}
\end{minipage}
\hfill
\begin{minipage}{0.45\textwidth}
\begin{center}
\includegraphics[scale=0.3]{images/clark_fast_vs_hco3.png}
\end{center}
\end{minipage}
\caption{Dependence of stable state in $CO_2$ and $HCO_3$ residing in carboxysome on the value of parameter $\mathit{fast}$ affecting the rate of carbon fixation reaction in the Clark model~\cite{clark14} studied in~\cite{benevs2018fully}.}\label{attractors}
\end{figure}

The core part of CMP is \emph{Biochemical Space} (BCS)~\cite{BCS}. BCS mechanistically describes the modelled process with emphasis placed on precise annotation by links to existing biological ontologies and by association with experimental data. This formal description then serves as a knowledge base for all mathematical models enabled by relating model components to BCS entities (for example of a BCS entity, see Figure~\ref{entity_example}). The mechanism of linking individual objects assigns biological meaning to mathematical constructs. This step is useful in the interpretation process of analysis results.

\begin{figure}[!h]
  \begin{center}
    \framebox{
      {\footnotesize
        \begin{tabular}{ l | l }
          ID: & KaiC\\
          Name: & circadian clock protein kinase KaiC\\
          Compartments: & cyt\\
          Composition: & S, T\\
          Classification: & protein\\
          Description: & Monomer component representing a core\\
          & component of the circadian clock system.\\
          Links: & UniProt::Q79PF4, CyanoBase::Synpcc7942\_1216\\
          Organism: & Synechococcus elongatus PCC 7942
      \end{tabular}}
    }
  \end{center}
  \caption{An example of complete information given for an agent representing circadian clock protein kinase KaiC of Synechococcus elongatus PCC 7942 strain. The emphasis is placed on the annotation -- relevant data entries are \href{https://www.uniprot.org/uniprot/Q79PF4}{Q79PF4} in UniProt database and \href{http://genome.microbedb.jp/cyanobase/GCA_000012525.1/genes/Synpcc7942_1216}{Synpcc7942\_1216} in CyanoBase. Links to other databases such as KEGG~\cite{KEGG} and ChEBI~\cite{ChEBI} are also supported.}\label{entity_example}
\end{figure}

An important feature of BCS is specification language it uses for the description of the processes. The key aspects necessary for such language are: human-readability -- easy to read, write, and maintain the description; executability -- defined semantics are needed for efficient consistency checking and analysis; universality -- multiple cellular processes with different level of detail can be sufficiently combined in a single specification; and syntactic scalability -- the combinatorial explosion of the description is avoided.

To that end, we have developed the first version of Biochemical Space Language~\cite{dved2016formal}. The language was presented in its basic form with emphasis on the annotation. The semantics of the language was defined indirectly by embedding to Kappa~\cite{kappa_formal}. However, BCSL aims at higher-level abstraction than Kappa that focuses on morphisms between protein binding sites. Therefore the Kappa-based formulation of BCSL has limited expressiveness and does not fit well the aims of our general modelling framework~\cite{klement2013comprehensive}. Additionally, Kappa does not provide a hierarchical description, which is one of the key aspects of BCSL.

In~\cite{trojak2018sasb}, we introduced a significantly improved version of BCSL with respect to the primary prototype. The new key aspects are: hierarchical and composable object types and rules are defined without the need to encode them in an existing rule-based framework thus avoiding any loss of information; executable semantics of rules is defined directly at the level of the language thus making a base for unique analysis tasks specific for the considered level of abstraction. Together with the formal definition, we provided a few static analysis techniques. Particularly, we showed how redundant rules in a model could be detected and automatically eliminated without loss of information, we provided a context-based reduction technique which produces an over-approximation of the modelled system, and we showed a static non-reachability technique which can check some reachability properties without the execution of model behaviour. On a case study of fibroblast growth factor signalling pathway, we demonstrated these techniques and showed how useful they could be in model building tasks.

The language is supported by BCSgen tool\footnote{\href{https://github.com/sybila/BCSgen}{https://github.com/sybila/BCSgen}}. The general goal of the tool is to provide the environment suitable for editing and maintenance of the BCSL models and provide the functionality to analyse the models. The tool was developed as a part of Dean's Program of the Faculty of Informatics MU for support of student research and development projects MUNI33/092015, where the only functionality provided by the tool was the import of the model from a text file (in a predefined format) and enumerating entire transition system of the model. The implementation was done by translating to Kappa and there was a simple graphical user interface.

The project MUNI33/062017 in the same program focused on the improvement of models maintenance, additional analysis services, and visualisation. The transition system generator was improved, using a reaction-based approach instead of translation to another language. Additionally, an interactive editor with automatic syntax highlighting and error messaging, import and export of the models, and reachability analysis were implemented. The graphical exploration of the transition system with highlighting the chosen node and announcing the content of the node and highlighting a path from the initial state to a state satisfying reachability condition were also implemented.

Currently, a web-based version eBCSgen\footnote{\href{https://github.com/sybila/eBCSgen}{https://github.com/sybila/eBCSgen}} is under development using Galaxy platform, which a first step towards integrating it within CMP.

\chapter{Publications}

\begin{itemize}

\item TrojÃ¡k, M., Å afrÃ¡nek, D., MertovÃ¡ L., Brim L.: Executable Biochemical Space for Specification and Analysis of Biochemical Systems. Submitted to Bioinformatics journal (2019)

(85\%) A journal paper presenting BCSL on an informal level. I formulated the language description and prepared case study demonstrating static analysis and usability of the language on different biological processes.

\item BeneÅ¡, N., Brim, L., Pastva, S., Å afrÃ¡nek, D., TrojÃ¡k, M., ÄervenÃ½, J., Å alagoviÄ, J.: Fully Automated Attractor Analysis of Cyanobacteria Models.. In: 2018 22nd International Conference on System Theory, Control and Computing (ICSTCC). pp. 354--359. IEEE (2018)

(14\%) A case study paper presenting the technique from~\cite{barnat2017detecting} applied to models from the e-cyanobacterium.org repository. I helped with the experimental evaluation and contributed to the text describing the models and the
evaluation results. The paper was awarded the best paper award in its category.

\item TrojÃ¡k, M., Å afrÃ¡nek, D., Brim, L., Å alagoviÄ, J., ÄervenÃ½, J.: Executable Biochemical Space for Specification and Analysis of Biochemical Systems. Accepted in 9th International Workshop on Static Analysis for Systems Biology (2018)

(80\%) A conference paper presenting a new version of BCSL. I formulated the formal definition of the language and static analysis, including their application examples.

\item TrojÃ¡k, M., Å afrÃ¡nek, D., Hrabec, J., Å alagoviÄ, J., RomanovskÃ¡, F., ÄervenÃ½, J.: E-cyanobacterium.org: A Web-based Platform for Systems Biology of Cyanobacteria. In: International Conference on Computational Methods in Systems Biology. pp. 316--322. Springer (2016)

(30\%) A tool paper presenting e-cyanobacterium.org as an instance of CMP. I helped with writing most of the text and brought several ideas to the web platform development.

\item DÄd, T., Å afrÃ¡nek, D., TrojÃ¡k, M., Klement, M., Å alagoviÄ, J., Brim, L.: Formal Biochemical Space with Semantics in Kappa and BNGL. Electronic Notes in Theoretical Computer Science 326, 27â49 (2016)

(30\%) A conference paper describing the first version of BCSL defined by embedding to another language. I helped to formulate the formal definition and algorithms for the embedding.

\end{itemize}

\newpage
\stepcounter{chapter}
\addcontentsline{toc}{chapter}{Bibliography}

\bibliographystyle{splncs03}
\bibliography{all}

% ------------------------------- appendix

\appendix

\chapter{Summary of Authorâs PhD Study}

\section{Posters and talks}

As a member of Systems Biology Laboratory and a PhD student, I had a few presentations on conferences:

\begin{itemize}
 \item TrojÃ¡k, M., Å afrÃ¡nek, D., Brim, L., Å alagoviÄ, J., ÄervenÃ½, J.: Executable biochemical space for specification and analysis of biochemical systems. In the 9th International Workshop on Static Analysis for Systems Biology (SASB 2018)
 \item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Å alagoviÄ, F. RomanovskÃ¡, and M. Hajnal: Biochemical Space: A Framework for Formal Description and Annotation of Complex Biological Processes. In Computational Modeling in Biology Network meeting (COMBINE 2017).
 \item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Å alagoviÄ, F. RomanovskÃ¡, and M. Hajnal: Biochemical Space: A Framework for Formal Description and Annotation of Complex Biological Processes. In the 8th International Workshop on Static Analysis for Systems Biology (SASB 2017).
\end{itemize}

\noindent and several poster presentations:

\begin{itemize}
	\item TrojÃ¡k, M., D. Å afrÃ¡nek, J. ÄervenÃ½, M. HavlÃ­k, L. MertovÃ¡, M. Hajnal, J. Hrabec, and J. Å alagoviÄ: Comprehensive Modelling Platform. In the 6th Workshop on Hybrid Systems \& Biology (HSB 2019).
	\item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Hrabec, J. Å alagoviÄ, J. ÄervenÃ½, M. Hajnal, L. MertovÃ¡, K. PalubovÃ¡, and M. HavlÃ­k: E-cyanobacterium.org: A Web-based Platform for Systems Biology of Cyanobacteria. In the 16th International Conference on Computational Methods in Systems Biology (CMSB 2018).
	\item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Hrabec, J. Å alagoviÄ, F. RomanovskÃ¡, M. Hajnal, and J. ÄervenÃ½: E-cyanobacterium.org: A Web-based Platform for Systems Biology of Cyanobacteria. In The Biomania Student Scientific Meeting 2017.
	\item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Hrabec, J. Å alagoviÄ, F. RomanovskÃ¡, M. Hajnal, and J. ÄervenÃ½: E-cyanobacterium.org: A Web-based Platform for Systems Biology of Cyanobacteria. In Doctoral Workshop on Mathematical and Engineering Methods in Computer Science 2017 (MEMICS 2017).
	\item TrojÃ¡k, M., D. Å afrÃ¡nek, J. Hrabec, J. Å alagoviÄ, F. RomanovskÃ¡, M. Hajnal, and J. ÄervenÃ½: E-cyanobacterium.org: A Web-based Platform for Systems Biology of Cyanobacteria. In Intelligent Systems for Molecular Biology and European Conference on Computational Biology (ISMB/ECCB 2017).
	\item TrojÃ¡k, M., T. DÄd, D. Å afrÃ¡nek, M. Klement, J. ÄervenÃ½, and F. RomanovskÃ¡: Biochemical Space: A Language for Formal Description and Annotation of Complex Biological Processes. In Doctoral Workshop on Mathematical and Engineering Methods in Computer Science 2016 (MEMICS 2016).
\end{itemize}

\section{Teaching, supervising, and reviewing}

I have tutored seminar groups on Foundations of Programming since autumn 2018. I have also been supervisor to one bachelor's thesis, \emph{Genome-wide extension of formal biochemical space} by KatarÃ­na PalubovÃ¡, reader to bachelor's thesis \emph{Boolean Network Analysis in Pithya Using Binary Decision Diagrams} by Jakub PolÃ¡Äek and bachelor's thesis \emph{Graphical User Interface for Visualisation of Simulations and Experiments in E-cyanobacterium.org} by Jozef MikuÅ¡inec, and reader to master's thesis \emph{Methods for approximative identification of Boolean networks} by Jaroslav Bayer and master's thesis \emph{Parallel Parameter Synthesis for Hybrid Systems} by Eva Å mijÃ¡kovÃ¡.

\chapter{Attached publications}

\includepdf[pages=-]{attached/sasb2015.pdf}

\includepdf[pages=-]{attached/sasb2018.pdf}

\end{document}

